src/openassetio-python/cmodule/src/errors/exceptionsBinding.cpp

```diff
@@ -462,3 +462,129 @@ void registerExceptions(const py::module &mod) {
 /**
  * @}
  */
+
+/**
+ * @section tocpp Conversion from Python exception to C++ exception.
+ *
+ * @{
+ */
+
+namespace {
+/**
+ * Hybrid of OpenAssetIO and pybind11 exception class.
+ *
+ * Multiple inheritance means that in C++ we can `catch` this exception
+ * as either `error_already_set` or the given OpenAssetIO exception. If
+ * (re)thrown, it can be caught further up the stack (again) as either
+ * of the two exceptions. In this way we can satisfy the two use-cases:
+ * - Catching the exception in C++ as an OpenAssetIO C++ exception.
+ * - Allowing the exception to propagate back to Python via pybind11,
+ *   which will translate it back into a Python exception (see
+ *   registerExceptions).
+ *
+ * Note that calling `.what()` on this exception is ambiguous and will
+ * cause a compiler error. However, this is a good thing, since this
+ * exception should never appear in a `catch` (should only be caught as
+ * one of the base classes).
+ *
+ * Generalisation assuming that the given exception type is a simple
+ * exception that takes a string message as its only constructor
+ * argument.
+ *
+ * @tparam CppException C++ exception type corresponding to given
+ * Python exception.
+ */
+template <class CppException>
+struct HybridException : py::error_already_set, CppException {
+  explicit HybridException(const py::error_already_set &pyExc)
+      : py::error_already_set{pyExc}, CppException{pyExc.what()} {}
+};
+
+/**
+ * Hybrid of OpenAssetIO and pybind11 exception class.
+ *
+ * Specialisation for more complex case of BatchElementException. See
+ * generalisation, above, for more details.
+ */
+template <>
+struct HybridException<BatchElementException> : py::error_already_set, BatchElementException {
+  explicit HybridException(const py::error_already_set &pyExc)
+      : py::error_already_set{pyExc},
+        BatchElementException{
+            py::cast<std::size_t>(pyExc.value().attr("index")),
+            py::cast<openassetio::errors::BatchElementError>(pyExc.value().attr("error")),
+            pyExc.what()} {}
+};
+
+/**
+ * Throw a HybridException if the given Python class name matches the
+ * given expected Python class name corresponding to the given C++
+ * exception type.
+ *
+ * @tparam Exception C++ exception to wrap in a HybridException.
+ * @param expectedPyExcName Python exception class name corresponding
+ * to @p Exception class.
+ * @param thrownPyExc pybind11-wrapped Python exception to potentially
+ * wrap in a HybridException.
+ * @param thrownPyExcName Python exception class name of @p thrownPyExc
+ * to test against @p expectedPyExcName.
+ */
+template <class Exception>
+void throwHybridExceptionIfMatches(const std::string_view expectedPyExcName,
+                                   const py::error_already_set &thrownPyExc,
+                                   const std::string_view thrownPyExcName) {
```

Comment:
[Consider] could you extract the name from the PyExc here in the function, rather than require an external extraction and argument pass??

Reply:
Could do, but it would be repeated multiple times (once for every call) rather than once in the caller and passed in.

Reply:
Isn't that already the case? Looks like `throwHybridExceptionIfMatches` is only called from inside `convertPyExceptionAndThrow`, and the extraction is being done each time that function is being called anyway.

Reply:
But `convertPyExceptionAndThrow` is only called once.

Reply:
Resolving this as it's not important, even though i still don't quite follow the argument... probably just text chat fog-of-war.